### DISCUSSION TITLE
elyra/kernel-py based off python 3.9

### DISCUSSION THREAD

**Discussion by amit-chandak-unskript** [jupyter/enterprise_gateway#1201] [Q&A]
Hi @kevin-bates any plans to support elyra kernel based off python 3.9. We use the native jupyterlab (without EG), which is based off 3.9, whereas the elyra kernel is still based off 3.7, causing us python pkg version manageability issues. 
Is it easy to build a kernel based off 3.9? I was not able to find the Dockerfile for the same. [This](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/contributors/docker.html#elyra-kernel-py) link talks its based off scipy-notebook which has moved to 3.10, hence the question if elyra also plans to move. I can help out if needed. Thanks

**Comment by kevin-bates** [comment]
Hi @amit-chandak-unskript - great question!

TL;DR - yes, let's update the tag of the `kernel-py` base image to `2022-01-24`.

---

Yes, we should update our version of python within the applicable images.  The Dockerfiles can be found in [`etc/docker`](https://github.com/jupyter-server/enterprise_gateway/tree/main/etc/docker) and, at a minimum, I think we should update the version of Python in [`kernel-py`](https://github.com/jupyter-server/enterprise_gateway/blob/2f9d635f9d9d85ae386b40f7b222b45ca580f978/etc/docker/kernel-py/Dockerfile#L2), [`kernel-tf-py`](https://github.com/jupyter-server/enterprise_gateway/blob/2f9d635f9d9d85ae386b40f7b222b45ca580f978/etc/docker/kernel-tf-py/Dockerfile#L3), [`kernel-tf-gpu-py`](https://github.com/jupyter-server/enterprise_gateway/blob/2f9d635f9d9d85ae386b40f7b222b45ca580f978/etc/docker/kernel-tf-gpu-py/Dockerfile#L2) and [`enterprise-gateway`](https://github.com/jupyter-server/enterprise_gateway/blob/2f9d635f9d9d85ae386b40f7b222b45ca580f978/etc/docker/enterprise-gateway/Dockerfile#L1).

As you've noticed, Python version updates are _inherited_, so we really need to update the corresponding base image.  Once done, a rebuild of `kernel-spark-py` (for example) will pickup the change made in `kernel-py`.

Of these images, all but `kernel-tf-gpu-py` derives from the [docker-stacks](https://github.com/jupyter/docker-stacks) images and I'd like to take this opportunity to convert to using the date-based tags for those.

I think we should try to move to Python 3.10, if that works for you. [*] (If not, I'm okay with 3.9, but a more current (yet stable) version is preferred.)  It looks like the docker-stacks images were [moved to 3.10](https://github.com/jupyter/docker-stacks/pull/1711) in late May - so you could probably look for date-based tags reflecting 2022-05-28 or more recent.

I suspect the tag on the base image for `kernel-tf-gpu-py` specifies the version of Python.  If we decide to only move to 3.9, then this would not require an update.

[*] Hmm - I see that the 2022-01-24 tagged images (used by `enterprise-gateway` and `kernel-tf-py`) are also already using Python 3.9, so now I'm inclined to simply update `kernel-py` to use `jupyter/scipy-notebook:2022-01-24` and be done.  We can look at moving to 3.10 later and this minimizes the change prior to a release.

Calling `docker build` within each directory is not sufficient to build the images since these are composed from various artifacts.  I would do the following:

```
make clean dist
make kernel-py kernel-spark-py
```
The inclusion of `kernel-spark-py` is just to ensure there aren't build issues introduced from the upgrade of `kernel-py`.

>  I can help out if needed.

Your help would be greatly appreciated!  (Sorry for the ramble)
